documentation/proposals/Proposal - Silk.NET 3.X Generic Maths.md

```diff
@@ -0,0 +1,884 @@
+    public static Vector2D<T> Hypot<T>(Vector2D<T> x, Vector2D<T> y) where T : IRootFunctions<T>;
+    public static Vector2D<T> Hypot<T>(Vector2D<T> x, T y) where T : IRootFunctions<T>;
+    public static Vector2D<T> RootN<T>(Vector2D<T> x, int n) where T : IRootFunctions<T>;
+
+    // IFloatingPoint<TSelf>
+    public static Vector2D<T> Round<T>(Vector2D<T> x) where T : IFloatingPoint<T>;
+    public static Vector2D<T> Round<T>(Vector2D<T> x, int digits) where T : IFloatingPoint<T>;
+    public static Vector2D<T> Round<T>(Vector2D<T> x, MidpointRounding mode) where T : IFloatingPoint<T>;
+    public static Vector2D<T> Round<T>(Vector2D<T> x, int digits, MidpointRounding mode) where T : IFloatingPoint<T>;
+    public static Vector2D<T> Truncate<T>(Vector2D<T> x) where T : IFloatingPoint<T>;
+
+    // IFloatingPointIeee754<TSelf>
+    public static Vector2D<T> Atan2<T>(Vector2D<T> x, Vector2D<T> y) where T : IFloatingPointIeee754<T>;
+    public static Vector2D<T> Atan2Pi<T>(Vector2D<T> x, Vector2D<T> y) where T : IFloatingPointIeee754<T>;
+    public static Vector2D<T> Atan2<T>(Vector2D<T> x, T y) where T : IFloatingPointIeee754<T>;
+    public static Vector2D<T> Atan2Pi<T>(Vector2D<T> x, T y) where T : IFloatingPointIeee754<T>;
+    public static Vector2D<T> BitDecrement<T>(Vector2D<T> x) where T : IFloatingPointIeee754<T>;
+    public static Vector2D<T> BitIncrement<T>(Vector2D<T> x) where T : IFloatingPointIeee754<T>;
+    public static Vector2D<T> FusedMultiplyAdd<T>(Vector2D<T> left, Vector2D<T> right, Vector2D<T> addend) where T : IFloatingPointIeee754<T>;
+    // public static Vector2D<T> Lerp<T>(Vector2D<T> value1, Vector2D<T> value2, Vector2D<T> amount) where T : IFloatingPointIeee754<T>;
+    public static Vector2D<T> ReciprocalEstimate<T>(Vector2D<T> x) where T : IFloatingPointIeee754<T>;
+    public static Vector2D<T> ReciprocalSqrtEstimate<T>(Vector2D<T> x) where T : IFloatingPointIeee754<T>;
+
+    // INumber<T>
+    // public static Vector2D<T> Clamp<T>(Vector2D<T> value, Vector2D<T> min, Vector2D<T> max) where T : INumber<T>;
+    public static Vector2D<T> CopySign<T>(Vector2D<T> value, Vector2D<T> sign) where T : INumber<T>;
+    public static Vector2D<T> CopySign<T>(Vector2D<T> value, T sign) where T : INumber<T>;
+    public static Vector2D<T> MaxNumber<T>(Vector2D<T> x, Vector2D<T> y) where T : INumber<T>;
+    public static Vector2D<T> MinNumber<T>(Vector2D<T> x, Vector2D<T> y) where T : INumber<T>;
+
+    // INumberBase<T>
+    // public static Vector2D<T> MaxMagnitude<T>(Vector2D<T> x, Vector2D<T> y) where T : INumberBase<T>;
+    // public static Vector2D<T> MaxMagnitudeNumber<T>(Vector2D<T> x, Vector2D<T> y) where T : INumberBase<T>;
+    // public static Vector2D<T> MinMagnitude<T>(Vector2D<T> x, Vector2D<T> y) where T : INumberBase<T>;
+    // public static Vector2D<T> MinMagnitudeNumber<T>(Vector2D<T> x, Vector2D<T> y) where T : INumberBase<T>;
+    // (there's no reason you would want these.)
```

Comment:
There are many reasons you would want these. Having a function that returns `|a|<|b| ? a : b` or `|a|>|b| ? a : b` is useful in a great many number of contexts.

One such example is accelerating various floating-point algorithms, especially where you need to do bitwise manipulation of the underlying representation. There are many other examples and general algorithms that can be accelerated using it as well, such as getting magnitude data for image/color processing or doing velocity field calculations.

Reply:
This was an oversight in writing the proposal - these methods are implemented elsewhere as they are required for INumberBase, and VectorXD<T> also provides Min/Max methods with NaN-ignoring behavior.